### PR TITLE
Update to OpenXR 1.1.36

### PR DESCRIPTION
**Pull request by rasmusgo**
This PR builds upon #157.

The biggest change in the generator is that `<feature>` tags are processed now. They are similar to `<extension>` but have a couple of important differences.

`<interaction_profiles>` were added in OpenXR 1.0.27 but we are still not using them. OpenXR 1.1.36 keeps building upon them with `<extend>` and `<interaction_profile>` tags inside `<require>` tags. I think we should consider adding support for them in another PR.

I used https://registry.khronos.org/OpenXR/specs/1.1/styleguide.html#schema and https://registry.khronos.org/vulkan/specs/1.3/registry.html to figure out the schema.

### CODE REVIEW

File: generator/src/main.rs [hunk 1]
```diff
@@ -162,6 +165,125 @@ impl Parser {
...
+                }
+                EndDocument => {
+                    panic!("unexpected end of document");
+                }
+                _ => {}
+            }
+        }
+    }
+
+    fn parse_feature_required(&mut self) {
```
Comment (Ralith) [inline generator/src/main.rs:195]: This looks like it has a lot of copy-paste from `parse_extension_required`. Can we factor that out?

**Review by Ralith — CHANGES_REQUESTED** [review]

File: generator/src/main.rs (hunk 1 above)
Comment (rasmusgo) [reply generator/src/main.rs]: I was able to unify them into `parse_required` by adding parameters to optionally pass extension name and number.

**Comment by rasmusgo** [comment]
- [ ] Links to `https://www.khronos.org/registry/OpenXR/specs/1.0/html/xrspec.html` should be updated to 1.1.

**Comment by awtterpip** [comment]
is there any ETA on this getting merged?

**Comment by Ralith** [comment]
It needs a rebase first, for one thing.

**Comment by rasmusgo** [comment]
Rebased now. The most glaring problem is that the instance cannot be created if OpenXR 1.1 isn't supported by the loader, as we try to use the current API version when creating the instance: https://github.com/Ralith/openxrs/blob/4681a25c910da1f078342935b60b4fabe3cc4747/openxr/src/entry.rs#L239

**Comment by Ralith** [comment]
Would it make sense to take a version argument, deferring that problem to the application? That'd help us gracefully handle movement of extensions that we provide high-level wrappers for into core, too.

File: generator/src/main.rs [hunk 2]
```diff
@@ -1308,13 +1317,10 @@ impl Parser {
     /// Generate high-level code
     #[allow(clippy::cognitive_complexity)] // TODO
     fn generate_hl(&self) -> TokenStream {
-        const BLACKLISTED_COMMANDS: [&str; 3] = [
-            "xrNegotiateLoaderRuntimeInterface",
-            "xrNegotiateLoaderApiLayerInterface",
-            "xrCreateApiLayerInstance",
-        ];
```
Comment (rasmusgo) [inline generator/src/main.rs:1268]: Keeping track of the commands listed in the feature tag for `XR_VERSION_1_0` allowed me to get rid of this! 😄

**Comment by rasmusgo** [comment]
This runs with api version 1.0 now. I'm not sure what, if anything, is missing from the 1.1 spec. The only thing that disappeared when when restricting the core commands to 1.0 was `xrLocateSpaces` from the high level bindings.

File: generator/src/main.rs (hunk 2 above)
Comment (Ralith) [reply generator/src/main.rs:1268]: Great catch!

File: generator/src/main.rs [hunk 3]
```diff
@@ -192,27 +192,112 @@ impl Parser {
...
+                        extensions: vec![ext],
+                    },
+                );
+            } else {
+                eprintln!("ignoring extension with unlisted tag: {}", ext_name);
+            }
+        }
+    }
+
+    fn parse_required(
```
Comment (Ralith) [inline generator/src/main.rs:267]: Can you squash this into the commit that introduces feature parsing? Or better yet, have an initial commit that factors out the shared code, followed by one that introduces feature parsing? That would make review much easier.

File: generator/src/main.rs [hunk 4]
```diff
@@ -1308,13 +1317,10 @@ impl Parser {
...
     #[allow(clippy::cognitive_complexity)] // TODO
     fn generate_hl(&self) -> TokenStream {
-        const BLACKLISTED_COMMANDS: [&str; 3] = [
-            "xrNegotiateLoaderRuntimeInterface",
-            "xrNegotiateLoaderApiLayerInterface",
-            "xrCreateApiLayerInstance",
-        ];
         let (instance_pfn_fields, instance_pfn_inits) = self.commands.iter().map(|(name, command)| {
-            if command.extension.is_some() || BLACKLISTED_COMMANDS.contains(&name.as_str()) {
+            if !command.feature.as_ref().map_or(false, |feature|
```
Comment (Ralith) [inline generator/src/main.rs:1321]: nit: push the `!` down for better readability

File: generator/src/main.rs [hunk 5]
```diff
@@ -1308,13 +1317,10 @@ impl Parser {
...
+                feature.as_ref() == "XR_VERSION_1_0"
```
Comment (Ralith) [inline generator/src/main.rs:1322]: Explain in a comment why we don't want 1.1 stuff here too.

**Review by Ralith — CHANGES_REQUESTED** [review]

File: generator/src/main.rs (hunk 3 above)
Comment (rasmusgo) [reply generator/src/main.rs:267]: Done!

File: generator/src/main.rs [hunk 6]
```diff
@@ -280,8 +313,7 @@ impl Parser {
                             } else if name.ends_with("EXTENSION_NAME") {
                                 let value = attr(&attributes, "value").unwrap();
                                 assert_eq!(&ext_name[..], &value[1..value.len() - 1]);
-                            } else {
-                                let value = attr(&attributes, "value").unwrap();
+                            } else if let Some(value) = attr(&attributes, "value") {
```
Comment (Ralith) [inline generator/src/main.rs]: Does removing the `unwrap` that was previously here put us at risk of missing future additions of new cases that the generator doesn't support?

File: generator/src/main.rs [hunk 7]
```diff
@@ -259,6 +292,12 @@ impl Parser {
                         if let Some(extends) = attr(&attributes, "extends") {
                             const EXT_BASE: i32 = 1_000_000_000;
                             const EXT_BLOCK_SIZE: i32 = 1000;
+                            let ext_number = ext_number.unwrap_or_else(|| {
+                                attr(&attributes, "extnumber")
```
Comment (Ralith) [inline generator/src/main.rs:304]: Are we relying on the caller to pass down a number for extensions, and parsing it out here instead for features? Would it be simpler/more consistent to either handle both cases (`number` vs. `extnumber`) automatically here, or alternatively to parse the number out uniformly in the caller, making the argument non-optional?

**Review by Ralith — APPROVED** [review]

File: generator/src/main.rs (hunk 6 above)
Comment (rasmusgo) [reply generator/src/main.rs]: Removing the unwrap should have come in a later commit but I missed to move it while rebasing.

File: generator/src/main.rs (hunk 7 above)
Comment (Ralith) [reply generator/src/main.rs:304]: Merging as-is for the sake of downstream.

**Pull request merged**